plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx

```diff
@@ -179,6 +186,14 @@ const productCollectionStore = {
 				yield prefetch( ref.href );
 			}
 		},
+		*onRender() {
+			const { collection } =
+				getContext< ProductCollectionStoreContext >();
+
+			triggerProductListRenderedEvent( {
+				collection: collection || CoreCollectionNames.PRODUCT_CATALOG,
+			} );
+		},
```

Comment:
### Major - Handle the scenario when Product Collection block isn't visible

I noticed that `wc-blocks_product_list_rendered` event is triggered even if Product Collection block isn't there on a post as you can see in screenshot below. It was surprising therefore I debugged it. I found out that there was "Top Rated" collection in Empty Mini Cart view, which is always rendered on the page. So, the event was triggered.

IMO, it's important to handle this scenario. One solution could be to verify in the `onRender` function whether the Product Collection block is visible in the viewport. If it's not visible, we shouldn't trigger the event.

What do you think? 

![image](https://github.com/user-attachments/assets/c20f41be-c619-4380-b56a-c669f0b45b3c)

![image](https://github.com/user-attachments/assets/8210d254-6648-4d8e-ab14-c922d84ee657)

Reply:
> One solution could be to verify in the onRender function whether the Product Collection block is visible in the viewport. If it's not visible, we shouldn't trigger the event.

But that means excluding many blocks from triggering the event. For example, the PC block rendered in the footer is not visible at first sight. And I don't think creating an observer to send an event when the block comes into viewport is the way we'd like to go with 🤔

Reply:
Oh, my bad! I used the wrong terminology, i.e., viewport 😅 I meant to say that we should check if the block is visible anywhere on the page, which means it doesn't have to be in the viewport. 

1. We will need to check if element is hidden using CSS i.e.
```tsx
function isVisible(element) {
  const style = window.getComputedStyle(element);
  return style.display !== 'none' && style.visibility !== 'hidden' && style.opacity !== '0';
}
````

2. We will need to check top, left, right, bottom of the element to see if it's visible on the page or not. Something like this:
```tsx
function isInViewport(element) {
  const rect = element.getBoundingClientRect();
  return (
    rect.top <= (window.innerHeight || document.documentElement.clientHeight) &&
    rect.left <= (window.innerWidth || document.documentElement.clientWidth) &&
    rect.bottom >= 0 &&
    rect.right >= 0
  );
}
```

2nd is the case with mini cart block which uses drawer. 

---

I will leave it to you to decide if we should handle this scenario or not. Because if we will decide to handle it then we can't use `data-wc-init` attribute to trigger the event, as it won't be re-triggered when we open the mini cart and the block becomes visible. We will probably need to use `data-wc-run`. However, this may result in more complex code. Therefore, I will leave it to you to decide if we should handle this scenario or not 🙂

Reply:
Interesting idea, @imanish003! My two cents here would be to leave this as it is and, if necessary, streamline it as you suggested.

I'm saying this because of potential third-party cases where the block might be hidden for some custom reason, and we cannot 100% control this. So, leaving it simple, I think, is best for now.

Reply:
I double-checked and previous blocks (All Products, Newest Products etc) trigger action on render no matter whether visible or not (for example if placed in Mini Cart). I'd keep it as-is, same as suggested by @xristos3490. Covering all of the edge cases would make it extremely complex IMO.

Reply:
@imanish003 with that decision, does it mean it could be merged?

Reply:
I am good with the decision to keep it as it is 🤝 

I think this can be merged after making a decision [here](https://github.com/woocommerce/woocommerce/pull/50166#discussion_r1714814545) 🙂